File: packages/client/src/topic.ts
import { MessageType } from "@conduit/shared";
import { EventEmitter } from "eventemitter3";

/** Events emitted by a {@link Topic} handle. */
export interface TopicEvents {
	/** Fired once the server confirms the subscription. */
	open: () => void;
	/** Fired for each publication matching this subscription. */
	message: (data: unknown, topic: string, from: string) => void;
	/** Fired when this peer has unsubscribed. */
	close: () => void;
	/** Fired when the server refuses an operation on this topic. */
	error: (error: Error) => void;
}

/** Sends a signaling message on behalf of a topic handle. */
export type TopicSend = (type: MessageType, payload: unknown) => void;

/**
 * A topic subscription.
 *
 * The pattern may be an exact topic name or a namespace prefix ending in `.*`;
 * the delivered `topic` is always the concrete name that was published, not the
 * pattern that matched it.
 */
export class Topic extends EventEmitter<TopicEvents> {
	private _open = false;
	private _closed = false;

	constructor(
		/** The subscribed topic name or prefix pattern. */
		readonly pattern: string,
		private readonly _send: TopicSend
	) {
		super();
	}

	/** Whether the server has confirmed the subscription. */
	get open(): boolean {
		return this._open;
	}

	/** Stop receiving publications on this subscription. */
	unsubscribe(): void {
		if (this._closed) {
			return;
		}
		this._send(MessageType.UNSUBSCRIBE, { topic: this.pattern });
	}

	// -- internal, driven by the owning Conduit -------------------------------

	/** @internal The server confirmed the subscription. */
	_confirm(): void {
		this._open = true;
		this.emit("open");
	}

	/** @internal A matching publication arrived. */
	_message(data: unknown, topic: string, from: string): void {
		this.emit("message", data, topic, from);
	}

	/** @internal The server refused an operation on this topic. */
	_error(message: string): void {
		this.emit("error", new Error(message));
	}

	/** @internal Tear the handle down, releasing listeners. */
	_close(): void {
		if (this._closed) {
			return;
		}
		this._closed = true;
		this._open = false;
		this.emit("close");
		this.removeAllListeners();
	}
}

File: packages/go-client/topic.go
package conduit

import (
	"strings"
	"sync"
)

// Topic is a handle to a topic subscription.
//
// The pattern may be an exact topic name or a namespace prefix ending in ".*".
// A Topic is safe for concurrent use.
type Topic struct {
	pattern string
	client  *Client

	mu     sync.RWMutex
	open   bool
	closed bool

	onMessage func(data any, topic string, from string)
	onClose   func()
}

func newTopic(pattern string, client *Client) *Topic {
	return &Topic{pattern: pattern, client: client}
}

// Pattern returns the subscribed topic name or prefix pattern.
func (t *Topic) Pattern() string {
	return t.pattern
}

// Open reports whether the server has confirmed the subscription.
func (t *Topic) Open() bool {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.open
}

// OnMessage sets the callback fired for each matching publication. The topic
// passed to it is the concrete published name, not the pattern that matched.
func (t *Topic) OnMessage(fn func(data any, topic string, from string)) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.onMessage = fn
}

// OnClose sets the callback fired once the subscription is removed.
func (t *Topic) OnClose(fn func()) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.onClose = fn
}

// Unsubscribe stops delivery on this subscription.
func (t *Topic) Unsubscribe() error {
	t.mu.RLock()
	closed := t.closed
	t.mu.RUnlock()
	if closed {
		return nil
	}

	msg, err := NewMessage(MessageTypeUnsubscribe, "", SubscribePayload{Topic: t.pattern})
	if err != nil {
		return err
	}
	return t.client.Send(msg)
}

// -- internal, driven by the owning Client ---------------------------------

func (t *Topic) confirm() {
	t.mu.Lock()
	if t.closed {
		t.mu.Unlock()
		return
	}
	t.open = true
	t.mu.Unlock()
}

func (t *Topic) message(data any, topic string, from string) {
	t.mu.RLock()
	if t.closed {
		t.mu.RUnlock()
		return
	}
	fn := t.onMessage
	t.mu.RUnlock()

	if fn != nil {
		fn(data, topic, from)
	}
}

func (t *Topic) close() {
	t.mu.Lock()
	if t.closed {
		t.mu.Unlock()
		return
	}
	t.closed = true
	t.open = false
	fn := t.onClose
	t.onMessage = nil
	t.onClose = nil
	t.mu.Unlock()

	if fn != nil {
		fn()
	}
}

// patternMatches reports whether a subscription pattern matches a published
// topic.
//
// Mirrors the server's rule: an exact name, or a namespace prefix ending in
// ".*". Applied client-side only to route one delivered copy to every matching
// handle; the server has already decided what this peer receives.
func patternMatches(pattern, topic string) bool {
	if pattern == topic {
		return true
	}
	if !strings.HasSuffix(pattern, ".*") {
		return false
	}
	// Keep the dot so "chat.*" matches "chat.general" but not "chatter.general".
	return strings.HasPrefix(topic, pattern[:len(pattern)-1])
}
